File: java/src/test/java/io/cucumber/gherkin/GherkinDocumentBuilderTest.java
package io.cucumber.gherkin;

import io.cucumber.messages.IdGenerator;
import io.cucumber.messages.types.Comment;
import io.cucumber.messages.types.FeatureChild;
import io.cucumber.messages.types.GherkinDocument;
import io.cucumber.messages.types.Location;
import io.cucumber.messages.types.Pickle;
import io.cucumber.messages.types.TableRow;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class GherkinDocumentBuilderTest {
    private final IdGenerator idGenerator = new IncrementingIdGenerator();

    @Test
    void is_reusable() {
        var parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        GherkinTokenMatcher matcher = new GherkinTokenMatcher();

        GherkinDocument d1 = parser.parse("Feature: 1", matcher, "1.feature");
        GherkinDocument d2 = parser.parse("Feature: 2", matcher, "2.feature");

        assertEquals("1", d1.getFeature().get().getName());
        assertEquals("2", d2.getFeature().get().getName());
    }

    @Test
    void parses_rules() {
        var parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        String data = """
                Feature: Some rules
                
                  Background:
                    Given fb
                
                  Rule: A
                    The rule A description
                
                    Background:
                      Given ab
                
                    Example: Example A
                      Given a
                
                  Rule: B
                    The rule B description
                
                    Example: Example B
                      Given b
                """;
        GherkinDocument doc = parser.parse(data, "test.feature");

        List<FeatureChild> children = doc.getFeature().get().getChildren();
        assertEquals(3, children.size());

        IdGenerator idGenerator = new IncrementingIdGenerator();
        PickleCompiler pickleCompiler = new PickleCompiler(idGenerator);
        List<Pickle> pickles = pickleCompiler.compile(doc, "hello.feature");
        assertEquals(2, pickles.size());

        assertEquals(3, pickles.get(0).getSteps().size());

        assertEquals(2, pickles.get(1).getSteps().size());

    }

    @Test
    void parses_just_comments() {
        var parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        GherkinDocument doc = parser.parse("# Just a comment", "test.feature");
        List<Comment> children = doc.getComments();
        assertEquals(1, children.size());
    }

    @Test
    void sets_empty_table_cells() {
        var parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));
        GherkinDocument doc = parser.parse("""
                        Feature:
                          Scenario:
                            Given a table
                              |a||b|
                        """,
                "test.feature"
        );
        List<FeatureChild> children = doc.getFeature().get().getChildren();
        assertEquals(1, children.size());
        TableRow row = children.get(0).getScenario().get().getSteps().get(0).getDataTable().get().getRows().get(0);
        assertEquals("a", row.getCells().get(0).getValue());
        assertEquals("", row.getCells().get(1).getValue());
        assertEquals("b", row.getCells().get(2).getValue());
    }

    @Test
    void table_cells_with_different_size_throws_exception() {
        // Given
        var parser = new Parser<>(new GherkinDocumentBuilder(idGenerator, "test.feature"));

        // When
        ParserException.CompositeParserException compositeParserException = assertThrows(ParserException.CompositeParserException.class, () -> parser.parse("""
                        Feature:
                          Scenario:
                            Given a table
                              |a|b|
                              |c|d|e|
                        """,
                "test.feature"
        ));

        // Then
        Assertions.assertThat(compositeParserException).hasMessageContaining("inconsistent cell count within the table");
        Assertions.assertThat(compositeParserException.errors)
                .singleElement()
                .extracting(e -> e.location)
                        .isEqualTo(new Location(5, 7));
    }

}

File: dart/test/src/ast/gherkin_document_builder_test.dart
import 'package:cucumber_gherkin/cucumber_gherkin.dart';
import 'package:cucumber_gherkin/src/ast/messages_gherkin_document_builder.dart';
import 'package:cucumber_gherkin/src/language/dialects_builtin.g.dart';
import 'package:cucumber_gherkin/src/language/gherkin_token_matcher.dart';
import 'package:cucumber_gherkin/src/language/string_token_scanner.dart';
import 'package:cucumber_gherkin/src/parser/parser.g.dart';
import 'package:cucumber_gherkin/src/pickles/messages_pickle_compiler.dart';
import 'package:cucumber_messages/cucumber_messages.dart' as messages;
import 'package:test/test.dart';

void main() {
  const languages = builtinDialects;
  final dialectProvider = GherkinDialectProvider(languages);
  final idGenerator = IdGenerator.incrementingGenerator;
  final builder = MessagesGherkinDocumentBuilder(idGenerator);
  final parser = Parser<messages.GherkinDocument>(builder);
  final matcher = GherkinTokenMatcher(dialectProvider);

  test('Is reusable', () {
    var tokenScanner = StringTokenScanner('Feature: 1');
    final d1 = parser.parse(tokenScanner, matcher);

    tokenScanner = StringTokenScanner('Feature: 2');
    final d2 = parser.parse(tokenScanner, matcher);

    expect(d1.feature!.name, '1');
    expect(d2.feature!.name, '2');
  });

  test('Parses rules', () {
    const data =
        'Feature: Some rules\n'
        '\n'
        '  Background:\n'
        '    Given fb\n'
        '\n'
        '  Rule: A\n'
        '    The rule A description\n'
        '\n'
        '    Background:\n'
        '      Given ab\n'
        '\n'
        '    Example: Example A\n'
        '      Given a\n'
        '\n'
        '  Rule: B\n'
        '    The rule B description\n'
        '\n'
        '    Example: Example B\n'
        '      Given b';

    final tokenScanner = StringTokenScanner(data);
    final doc = parser.parse(tokenScanner, matcher);

    final children = doc.feature!.children.toList();
    expect(children.length, 3);

    final pickleCompiler = MessagesPickleCompiler(idGenerator);
    final pickles = pickleCompiler.compile(doc, 'hello.feature');
    expect(2, pickles.length);

    expect(3, pickles.first.steps.length);

    expect(2, pickles.elementAt(1).steps.length);
  });

  test('Parses just comments', () {
    const data = '# Just a comment';
    final tokenScanner = StringTokenScanner(data);
    final doc = parser.parse(tokenScanner, matcher);
    final children = doc.comments;
    expect(children.length, 1);
  });

  test('Sets empty table cells', () {
    const data =
        'Feature:\n'
        '  Scenario:\n'
        '    Given a table\n'
        '      |a||b|';
    final tokenScanner = StringTokenScanner(data);
    final doc = parser.parse(tokenScanner, matcher);

    final row =
        doc.feature!.children.first.scenario!.steps.first.dataTable!.rows.first;
    expect('a', row.cells.first.value);
    expect('', row.cells.elementAt(1).value);
    expect('b', row.cells.elementAt(2).value);
  });
}
